source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,582 @@
+ *
+ * */
+static uint32_t prvAzureIoTHubClientDeviceTwinProcess( AzureIoTHubClientReceiveContext_t * pxContext,
+                                                       AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                       void * pxPublishInfo )
+{
+    AzureIoTHubClientTwinResponse_t xTwinResponse;
+    AzureIoTMQTTPublishInfo_t * xMQTTPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pxPublishInfo;
+    az_iot_hub_client_twin_response xOutRequest;
+    az_span xTopicSpan = az_span_create( ( uint8_t * ) xMQTTPublishInfo->pTopicName, xMQTTPublishInfo->topicNameLength );
+    uint32_t xRequestID = 0;
+
+    /* Failed means no topic match */
+    if( az_result_failed( az_iot_hub_client_twin_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                       xTopicSpan, &xOutRequest ) ) )
+    {
+        return eAzureIoTHubClientTopicNoMatch;
+    }
+
+    /* Process incoming Publish. */
+    AZLogInfo( ( "Twin Incoming QoS : %d.", xMQTTPublishInfo->qos ) );
+
+    /* Verify the received publish is for the feature we have subscribed to. */
+    AZLogInfo( ( "Twin topic: %.*s. with payload : %.*s.",
+                 xMQTTPublishInfo->topicNameLength,
+                 xMQTTPublishInfo->pTopicName,
+                 xMQTTPublishInfo->payloadLength,
+                 xMQTTPublishInfo->pPayload ) );
+
+    if( pxContext->_internal.callbacks.xTwinCallback )
+    {
+        if( az_span_size( xOutRequest.xRequestID ) == 0 )
+        {
+            xTwinResponse.messageType = eAzureIoTHubTwinDesiredPropertyMessage;
+        }
+        else
+        {
+            if( az_result_failed( az_span_atou32( xOutRequest.xRequestID, &xRequestID ) ) )
+            {
+                return eAzureIoTHubClientFailed;
+            }
+
+            if( xRequestID & 0x01 )
+            {
+                xTwinResponse.messageType = eAzureIoTHubTwinReportedResponseMessage;
+            }
+            else
+            {
+                xTwinResponse.messageType = eAzureIoTHubTwinGetMessage;
+            }
+        }
+
+        xTwinResponse.pvMessagePayload = xMQTTPublishInfo->pPayload;
+        xTwinResponse.ulPayloadLength = ( uint32_t ) xMQTTPublishInfo->payloadLength;
+        xTwinResponse.messageStatus = xOutRequest.status;
+        xTwinResponse.pucVersion = az_span_ptr( xOutRequest.version );
+        xTwinResponse.usVersionLength = ( uint16_t ) az_span_size( xOutRequest.version );
+        xTwinResponse.ulRequestID = xRequestID;
+        pxContext->_internal.callbacks.xTwinCallback( &xTwinResponse,
+                                                      pxContext->_internal.pvCallbackContext );
+    }
+
+    return eAzureIoTHubClientSuccess;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Time callback for MQTT initialization.
+ *
+ * */
+static uint32_t prvGetTimeMs( void )
```

Comment:
>void [](start = 30, length = 4)

do we do this? I just want to be consistent

Reply:
I think we should for functions with no args